### PR TITLE
AMP-49662 React Native example (TypeScript)

### PR DESCRIPTION
**Pull request by falconandy**
1. did not test on iOS
2. no luck to setup unit tests yet => no github action

### CODE REVIEW

File: react-native/typescript/AmpliApp/middleware/segmentMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ */
+export function getSegmentMiddleware(config: Config): Middleware {
+  // Create Segment tracker
+  const analytics = createClient(config);
+
+  // Create Segment Middleware
+  return (payload, next) => {
+    const {event, extra} = payload;
+    const segmentExtra = extra as SegmentExtra;
+    const userId = segmentExtra?.segment?.userId;
```
Comment (justin-fiedler) [inline react-native/typescript/AmpliApp/middleware/segmentMiddleware.ts:28]: I think `userId` might be confusing here but don't see `anonymousId` in the Segment RN SDK. Can we use another segment specific option like `integrations`? https://segment.com/docs/connections/sources/catalog/libraries/mobile/react-native/#block-specific-events-from-going-to-a-destination

**Review by justin-fiedler — CHANGES_REQUESTED** [review]

File: react-native/typescript/AmpliApp/App.tsx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    <View>
+      <View style={styles.appHeader}>
+        <Button
+          title="Identify"
+          onPress={() => ampli.identify(userId, {requiredNumber: 42})}
+        />
+
+        <Button
+          title="Set Group"
+          onPress={() => ampli.setGroup('test group', 'browser-ts-ampli')}
```
[the remark this answers is not in the document]
Comment (falconandy) [reply react-native/typescript/AmpliApp/App.tsx]: Fixed

File: react-native/typescript/AmpliApp/App.tsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        />
+
+        <Button
+          title="Group Identify"
+          onPress={() =>
+            ampli.groupIdentify('test group', 'browser-ts-ampli', {
```
[the remark this answers is not in the document]
Comment (falconandy) [reply react-native/typescript/AmpliApp/App.tsx]: Fixed

File: react-native/typescript/AmpliApp/middleware/segmentMiddleware.ts (hunk 1 above)
Comment (falconandy) [reply react-native/typescript/AmpliApp/middleware/segmentMiddleware.ts]: It looks the current version of Segment RN SDK (2+) doesn't support any additional track options:
```
track(eventName: string, options?: JsonMap)
```
https://github.com/segmentio/analytics-react-native/blob/master/packages/core/src/analytics.ts#L484

I use `2.1.7` version - maybe it is beta?
```
"@segment/analytics-react-native": "^2.1.7"
```

Should I use stable `1.x.y` version instead?

BTW, I changed `userId` logic

File: react-native/typescript/AmpliApp/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+ * @format
+ */
+
+import React from 'react';
+import {Button, StyleSheet, View} from 'react-native';
+import Config from 'react-native-config';
+
+import {ampli} from './ampli';
+import {EventWithOptionalProperties} from './ampli';
+// import {getSegmentMiddleware} from './middleware/segmentMiddleware';
```
[the remark this answers is not in the document]
Comment (falconandy) [reply react-native/typescript/AmpliApp/src/App.tsx:17]: It is commented out to hide eslint warning `ESLint: 'getSegmentMiddleware' is defined but never used.(@typescript-eslint/no-unused-vars)`

**Comment by falconandy** [comment]
>  no luck to setup unit tests yet => no github action

Done

File: react-native/typescript/AmpliApp/src/App.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (justin-fiedler) [reply react-native/typescript/AmpliApp/src/App.tsx:17]: To make it easier for users to toggle middleware on/off can you please add a top level `/* eslint-disable @typescript-eslint/no-unused-vars */` instead.

File: react-native/typescript/AmpliApp/src/middleware/loggingMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * Logs all the things
+ *
+ * @param payload
+ * @param next
+ */
+export const loggingMiddleware: Middleware = (payload, next) => {
+  const jsons = (o: any) => JSON.stringify(o, undefined, 2);
+
+  const {event} = payload as any;
```
Comment (justin-fiedler) [inline react-native/typescript/AmpliApp/src/middleware/loggingMiddleware.ts:14]: Why `payload as any` here?

File: react-native/typescript/AmpliApp/src/middleware/loggingMiddleware.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+  const eventJson = jsons(event);
+
+  console.log(`[ampli] event='${eventJson}`);
```
Comment (justin-fiedler) [inline react-native/typescript/AmpliApp/src/middleware/loggingMiddleware.ts:18]: Can you also output the `extra`

File: react-native/typescript/AmpliApp/src/middleware/segmentMiddleware.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  Middleware,
+  SpecialEventType,
+} from '@amplitude/react-native';
+
+/**
+ * getSegmentMiddleware
+ *
+ * Initializes Segment tracker and returns Segment Middleware
+ */
+export function getSegmentMiddleware(config: Config): {
```
Comment (justin-fiedler) [inline react-native/typescript/AmpliApp/src/middleware/segmentMiddleware.ts:16]: Thanks for this!

**Review by justin-fiedler — CHANGES_REQUESTED** [review]
Overall looks great. See minor comments.

File: react-native/typescript/AmpliApp/middleware/segmentMiddleware.ts (hunk 1 above)
Comment (justin-fiedler) [reply react-native/typescript/AmpliApp/middleware/segmentMiddleware.ts]: Current implementation with 2.1.7 looks good. Thanks!

File: react-native/typescript/AmpliApp/src/App.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (falconandy) [reply react-native/typescript/AmpliApp/src/App.tsx]: Fixed

File: react-native/typescript/AmpliApp/src/middleware/loggingMiddleware.ts (hunk 5 above)
Comment (falconandy) [reply react-native/typescript/AmpliApp/src/middleware/loggingMiddleware.ts]: Fixed

File: react-native/typescript/AmpliApp/src/middleware/loggingMiddleware.ts (hunk 6 above)
Comment (falconandy) [reply react-native/typescript/AmpliApp/src/middleware/loggingMiddleware.ts]: Done

**Review by justin-fiedler — APPROVED** [review]

**Pull request merged**